### DISCUSSION TITLE
Link Margin Issue

### DISCUSSION THREAD

**Discussion by kmfnjtze8888** [openthread/openthread#8642] [Q&A]
Hello,

I work for a company that uses OpenThread for our production units to communicate in the field. We are using Silicon Labs' radio boards with the EFR32MG12 and EFR32BG12 chips, and we use 915 MHz frequency. 

Currently, we are having issues where some of our units are choosing neighbors with low RSSI values (e.g., below -90).

Last year, we had an issue with our devices reporting incorrect RSSI values, and I created a discussion in the OpenThread forum (link below). @schinchani from Silicon Labs helped us resolve this issue by suggesting to change two lines in the PHY code, which fixed the issue (see link below).

https://github.com/openthread/openthread/discussions/7750

However, we suspect that this issue is not yet fixed when calculating the link margin. I have noticed that for one of our border routers, it somehow chooses devices with poor RSSI values (i.e., below -90) as its neighbors. I created a discussion on the OpenThread forum to inquire about this and I was told by one of the maintainers that "Routers do not attempt to establish links with neighboring routers when the link margin is below threshold" (see link below).

https://github.com/openthread/openthread/discussions/8636

As far as I know, the Link Margin is calculated by the the difference between the RSSI and radio's receive sensitivity (i.e., 100 dBm), and we never changed the minimum MLE link request margin when we built the project, so the min is the default value (i.e., 10). Therefore, this means that the border router should only form a link with devices that have RSSI of -90 dBm or higher.

However, for some reason our border router chose three units with RSSI values below -90 as its neighbors, and using the command "scan" to scan for all the devices around this border router shows that those 3 neighbors have the worst RSSI values compared to the rest in this network.

![image](https://user-images.githubusercontent.com/72487638/211663822-22e891ce-e7b8-4f96-9685-05b002bf2bb6.png)

We suspect that the RSSI values that are being used to calculate the link margin still has that positive RSSI issue that we initially encountered last year, which seems to happen when the signal is weak, and are not the same as those we see in "neighbor table" or "scan". If they are, then that would mean that any device with poor RSSI values will have a link margin of 100+ which then prompts OpenThread to choose those as the neighbors.

Can someone point me in the right direction on how to resolve this? We have thousands of units that currently use these radios in the field so it is quite critical for us. Any help is appreciated. Thanks!

**Comment by Irving-cl** [comment]
Hi @kmfnjtze8888,

I think we can add more logs to see how the LinkMargin were actually calcuated for the 3 neighbors and see what's wrong.

I made a [commit](https://github.com/openthread/openthread/compare/main...Irving-cl:openthread:debug_8642) for the logs to add. Could you re-build your BR with the logs I added and then reproduce the issue again? After that, we can match the Extended MAC address and see how LinkMargin for those neighbors were calculated. And let's see if there are issues with RSSI or Noise Floor.

Make sure you turned logs on when building the BR.

**Reply by kmfnjtze8888** [reply]
Hi @Irving-cl,

I rebuilt my RCP with the logs you added and also enabled the all logging by setting `#define OPENTHREAD_FULL_LOGS_ENABLE 1` in **sl_openthread_features_config.h**. I was able to reproduce the issue but I do not see any logs.

I also tried adding the logs to the BR host raspberry pi, but then the OT stops working (i.e., `sudo ot-ctl` won't work)

Do you mind pointing me to the right direction on what I'm doing wrong here? Thanks.

**Comment by Irving-cl** [comment]
Hi @kmfnjtze8888,

The logs should be added on the host side instead of on RCP. We don't need to update the RCP.

I think logs are by default enabled on otbr and now dynamic log level is by default. We can check log level by `ot-ctl log level` and set it by `ot-ctl log level <1-5>`.

> I also tried adding the logs to the BR host raspberry pi, but then the OT stops working (i.e., sudo ot-ctl won't work)

This basically means otbr-agent is not running. Could you check the status of otbr-agent by `sudo service otbr-agent status` on the pi?

**Reply by kmfnjtze8888** [reply]
> This basically means otbr-agent is not running. Could you check the status of otbr-agent by sudo service otbr-agent status on the pi?

It is working now and I added the logs on the host and re-built it. I am working on an older commit from last year 68c69910d0a13096c2ed804310c65a0888fd3059 because it what we currently use in production so **mle_router.cpp** is very different, but I was able to find the equivalent methods on where to put the logs for lines 1144 and 1152, but not for 1451 and 1466.

I re-created the issue and from my tests, the rss and the calculated link margins were correct, and also the noise floor was the expected value which was the radio's receive sensitivity (-100 dB). However, the border router still picks devices with rssi values below -90 as neighbors. From the mle_router.cpp, it looks like routers shouldn't form a link with each other when the link margin is less than 10 (default - and I didn't change this default value), and the calculated link margins from what I've seen were correct but still somehow picks those with rssi lower than -90 as neighbors. Do you know where I can find where neighbors are created outside of mle_router.cpp?

**Reply by jwhui** [reply]
MLE is the only component responsible for establishing Thread neighbors.

Note that RSSI is time varying and can fluctuate from packet to packet. The 10 dBm threshold is a hysteresis to account for the variation. Note that sensitivity represents the minimum RSSI at which the radio can continue to receive packets. Make sure your device can receive reliably at -100 dBm when there is no external noise.

**Reply by kmfnjtze8888** [reply]
Isn't mle_router.cpp for establishing links between routers or leaders? Where do I find the ones for child or just establishing links in general? Is it in mle.cpp?

**Reply by jwhui** [reply]
`mle.cpp` handles links between parents and children, but that occurs using different messages (Parent Request/Response, Child ID Request/Response) and is not subject to the 10 dBm hysteresis.
